File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.py
import sys
sys.setrecursionlimit(10**6)

def solution(k, room_number):
    def check(num):
        if num not in reserved.keys():
            reserved[num] = num + 1
            return num
        
        empty = check(reserved[num])
        reserved[num] = empty + 1
        return empty
        
        
    answer = []    
    reserved = dict()
    for number in room_number:
        answer.append(check(number))
    
    return answer

File: 프로그래머스/4/64063. 호텔 방 배정/호텔 방 배정.java
import java.util.*;

class Solution {
        public long[] solution(long k, long[] room_number) {
            Map<Long, Long> rooms = new HashMap<>();
            long[] answer = new long[room_number.length];
            for (int i = 0; i < room_number.length; i++) {
                answer[i] = findEmptyRoom(room_number[i], rooms);
            }
            return answer;
        }

        public long findEmptyRoom(long number, Map<Long, Long> rooms) {
            if (!rooms.containsKey(number)) {
                rooms.put(number, number + 1);
                return number;
            }

            long next = findEmptyRoom(rooms.get(number), rooms);
            rooms.replace(number, next + 1);
            return next;
        }

    }